modules/fedimint-mint-tests/src/bin/mint-client-restore.rs

```diff
@@ -0,0 +1,91 @@
+use anyhow::Result;
+use devimint::cmd;
+use devimint::federation::Federation;
+use devimint::util::FedimintCli;
+use devimint::version_constants::VERSION_0_5_0_ALPHA;
+use fedimint_logging::LOG_DEVIMINT;
+use rand::Rng;
+use tracing::{debug, info};
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    devimint::run_devfed_test(|fed, _process_mgr| async move {
+        let fed = fed.fed().await?;
+
+        test_restore_gap_test(fed).await?;
+        Ok(())
+    })
+    .await
+}
+
+pub async fn test_restore_gap_test(fed: &Federation) -> Result<()> {
```

Comment:
It says something about the gap limit, but it's unclear to me which guarantees about preconditions the code provides. The coin selection and reissuance is opaque, so we don't really know which scenario will be tested. Ideally we could test scenarios where transactions failed and e-cash outputs were cancelled, producing holes in the used indexes.

Reply:
@elsirion So your worry is that on failed paymenets client allocates the note_idx, but never uses it, potentially creating holes of unused mode_idx larger than gap limit? In that case why don't we just increase the gap limit? I mean - there is no other way anyway.

Reply:
Yeah, that came to my mind too. I've been staring at the mint recovery code all afternoon and haven't seen anything suspicious.

Reply:
@dpc https://github.com/fedimint/fedimint/pull/6096

Reply:
@elsirion Sir, I can test that to, see last commit.

Reply:
I think we've found something.

Reply:
@elsirion Actually, it was just artifact of multiple gaps adding up and crossing the gap limit, it seems. Everything seems to be working.